Bug 1281208 - Aggregate numeric scalars

fbertsch opened the pull request:
This does not include numeric scalars in child payloads.

@vitillo r?

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/mozilla/python_mozaggregator/22)
<!-- Reviewable:end -->

vitillo commented:
Please increase the version number.

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[mozaggregator/aggregator.py, line 16 at r1](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-KXGsu3UeUzE0rMV9bn3:-KXGsu3UeUzE0rMV9bn4:bcsmarm) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/537e59b3b5ca0806040a4196bffda46a87f809fa/mozaggregator/aggregator.py#L16)):*
> ```Python
> # Simple measurement and count histogram labels
> simple_measures_labels = cached_exponential_buckets(1, 30000, 50)
> numeric_scalars_labels = cached_exponential_buckets(1, 30000, 50)
> ```

I am not so sure that this is the right set of labels. Initially we wanted to use these labels as well for the count histograms but they turned out to have a bad accuracy at the beginning of the interval, which is where it was most needed. It would be helpful to run a quick analysis to verify how the proposed bucketing affects the scalars we currently have. I suspect we might want to treat them similarly to count histograms.

---

*[mozaggregator/aggregator.py, line 192 at r1](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-KXGttJrxjANlqSwFk7k:-KXGttJrxjANlqSwFk7l:b8vtgzd) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/537e59b3b5ca0806040a4196bffda46a87f809fa/mozaggregator/aggregator.py#L192)):*
> ```Python
> 
> def _extract_numeric_scalars(state, scalar_dict, is_child=False):
>     return _extract_scalar_values(state, scalar_dict, "NUMERIC_SCALAR", numeric_scalars_labels, is_child)
> ```

What about using just  "SCALAR" as prefix?

---

*[tests/test_aggregator.py, line 96 at r1](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-KXGvYmIjqBchdJBroA-:-KXGvYmJeFSi1zS4teHg:b-2ki032) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/537e59b3b5ca0806040a4196bffda46a87f809fa/tests/test_aggregator.py#L96)):*
> ```Python
>                 metric_count[metric] += 1
>                 assert(label == "")
>                 assert(value["count"] == (NUM_PINGS_PER_DIMENSIONS - NUM_AGGREGATED_CHILD_PINGS))
> ```

Can we use `expected_count()`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-:-KXGw-xZYvztrMUQ9T7J:b-d4b8nz)*
<!-- Sent from Reviewable.io -->

fbertsch commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[mozaggregator/aggregator.py, line 16 at r1](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-KXGsu3UeUzE0rMV9bn3:-KXIFrJoS2lFFWwKGM3r:b-ybd6lm) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/537e59b3b5ca0806040a4196bffda46a87f809fa/mozaggregator/aggregator.py#L16)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary>
> I am not so sure that this is the right set of labels. Initially we wanted to use these labels as well for the count histograms but they turned out to have a bad accuracy at the beginning of the interval, which is where it was most needed. It would be helpful to run a quick analysis to verify how the proposed bucketing affects the scalars we currently have. I suspect we might want to treat them similarly to count histograms.
</details>
So I looked at .5% of all data for a day and here's what I found: Average error would be .984 (value - bucket_label).  81% of the data is 0 <= x < 10, 90% 0 <= x < 20. 

Based on these we're not really losing too much accuracy, but I agree that at the very least for x <= 10 we should have labels and values match.

---

*[mozaggregator/aggregator.py, line 192 at r1](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-KXGttJrxjANlqSwFk7k:-KXIJZmSB81D0cdqZber:b-896fix) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/537e59b3b5ca0806040a4196bffda46a87f809fa/mozaggregator/aggregator.py#L192)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary>
> What about using just  "SCALAR" as prefix?
</details>
Done.

---

*[tests/test_aggregator.py, line 96 at r1](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-KXGvYmIjqBchdJBroA-:-KXIKvIcza3XMDTdApVW:b-n4z9sg) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/537e59b3b5ca0806040a4196bffda46a87f809fa/tests/test_aggregator.py#L96)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary>
> Can we use `expected_count()`?
</details>
expected_count() would return NUM_PINGS_PER_DIMENSIONS.

If you think it should be used here, perhaps I put the data in the wrong place in generate_payload(). Thoughts?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22)*
<!-- Sent from Reviewable.io -->

vitillo commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[mozaggregator/aggregator.py, line 16 at r1](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-KXGsu3UeUzE0rMV9bn3:-KXLcNDk4A32zk-dthsn:b-jfbrv2) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/537e59b3b5ca0806040a4196bffda46a87f809fa/mozaggregator/aggregator.py#L16)):*
<details><summary><i>Previously, fbertsch (Frank Bertsch) wrote…</i></summary>
> So I looked at .5% of all data for a day and here's what I found: Average error would be .984 (value - bucket_label).  81% of the data is 0 <= x < 10, 90% 0 <= x < 20. 
> 
> Based on these we're not really losing too much accuracy, but I agree that at the very least for x <= 10 we should have labels and values match.
</details>
Let's use the same labels as count histograms then.

---

*[tests/test_aggregator.py, line 93 at r1](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-KXLcbNYB4R9NzY5eLju:-KXLcbN_VAPgGW1feWBu:b-rhvhn5) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/537e59b3b5ca0806040a4196bffda46a87f809fa/tests/test_aggregator.py#L93)):*
> ```Python
>             metric, label, child = key
> 
>             if metric.startswith("NUMERIC_SCALAR_"):
> ```

How is this even working now that we are using the "SCALAR" prefix?

---

*[tests/test_aggregator.py, line 96 at r1](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-KXGvYmIjqBchdJBroA-:-KXLchoHTVFtD4Insm_k:bkpd1au) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/537e59b3b5ca0806040a4196bffda46a87f809fa/tests/test_aggregator.py#L96)):*
<details><summary><i>Previously, fbertsch (Frank Bertsch) wrote…</i></summary>
> expected_count() would return NUM_PINGS_PER_DIMENSIONS.
> 
> If you think it should be used here, perhaps I put the data in the wrong place in generate_payload(). Thoughts?
</details>
I would like to understand why we can't use the same code as for count histograms. It's probably worthwhile leaving a comment on why the differ, if they differ.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22)*
<!-- Sent from Reviewable.io -->

fbertsch commented:
Review status: 1 of 4 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[mozaggregator/aggregator.py, line 16 at r1](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-KXGsu3UeUzE0rMV9bn3:-KXgs2-kkRgsaxp44AA2:b-896fix) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/537e59b3b5ca0806040a4196bffda46a87f809fa/mozaggregator/aggregator.py#L16)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

Let's use the same labels as count histograms then.
</blockquote></details>

Done.

---

*[tests/test_aggregator.py, line 93 at r1](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-KXLcbNYB4R9NzY5eLju:-KXgs6WiHANMv2jMNYVB:bfjyrzi) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/537e59b3b5ca0806040a4196bffda46a87f809fa/tests/test_aggregator.py#L93)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

How is this even working now that we are using the "SCALAR" prefix?
</blockquote></details>

It wasn't. Fixed.

---

*[tests/test_aggregator.py, line 96 at r1](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-KXGvYmIjqBchdJBroA-:-KXgsBNnd2pXrv_KnXoo:b-tqykne) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/537e59b3b5ca0806040a4196bffda46a87f809fa/tests/test_aggregator.py#L96)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

I would like to understand why we can't use the same code as for count histograms. It's probably worthwhile leaving a comment on why the differ, if they differ.
</blockquote></details>

I looked into it some more. I did have the creation in the wrong place in generate_payload(). Now it uses expected_count().

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22)*
<!-- Sent from Reviewable.io -->

fbertsch commented:
@vitillo RFAL

vitillo commented:
Please rebase your PR.

fbertsch commented:
@vitillo rebased.

fbertsch commented:
@vitillo, this is RFAL.

vitillo commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-:-KZR_VARM4pfg6Pe08rI:bnfp4nl)*
<!-- Sent from Reviewable.io -->

vitillo commented:
@fbertsch Please add a test in `test_db.py` as well. How is the dashboard going to visualize these histograms? Is there a PR open for that?

fbertsch commented:
@vitillo: since I used the count histogram buckets, they will be visualized exactly as those are, correct? It should be no different than adding another probe. I can add tests for db.py.

vitillo commented:
Right, but will it work out of the box without any changes to the dashboard?

fbertsch commented:
@vitillo: Code has been updated with tests in `test_db.py`, which of course required other changes to the service and `db.py` (thanks for noticing that!). 
Re: TMO, the only part that *may* be missing is the metric names showing up in the list. I'll check to see if changes are required for that.

vitillo commented:
OK, please let me know when you tested it with the dashboard.

fbertsch commented:
@vitillo, everything is showing up as expected on TMO: https://pageshot.net/images/287f8f93-2747-4882-8b00-533ea0260948.png

I think we are good to go with this change.

vitillo commented:
Reviewed 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[mozaggregator/db.py, line 99 at r4](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_DZgNiNjzajRUrH8b2:-K_DZgNiNjzajRUrH8b3:b552gd9) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/6c8f76af4b3321bedfef4079f1ef6029c37dbb32/mozaggregator/db.py#L99)):*
> ```Python
>             histogram = pd.Series({int(k): v for k, v in values.iteritems()}, index=labels).fillna(0).values
>             break
>     else:
> ```

Which condition does this `else` clause belong to? 

---

*[tests/dataset.py, line 112 at r4](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_DeSj2aQXhH74TZM_h:-K_DeSj2aQXhH74TZM_i:b-muvcq3) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/6c8f76af4b3321bedfef4079f1ef6029c37dbb32/tests/dataset.py#L112)):*
> ```Python
> simple_measurements_template = {"uptime": SCALAR_VALUE, "addonManager": {u'XPIDB_parseDB_MS': SCALAR_VALUE}}
> 
> scalars_template = {"browser-engagement-total_uri_count": SCALAR_VALUE, "browser-engagement": {"test": SCALAR_VALUE}}
> ```

Are you sure this is the right format? I don't see nested scalars in the raw pings. Maybe @dexterp37 can weigh in. It would also be useful to put a reference to the scalars format in a comment. 

---

*[tests/test_db.py, line 245 at r4](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_Dg9QdA_XnyPLlDFhs:-K_Dg9QdA_XnyPLlDFht:b9mdiwj) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/6c8f76af4b3321bedfef4079f1ef6029c37dbb32/tests/test_db.py#L245)):*
> ```Python
> 
> @nottest
> def test_numeric_scalar(prefix, channel, version, dates, metric, value, expected_count):
> ```

Can `test_simple_measure` and `test_numeric_scalar` be refactored into the same function?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22)*
<!-- Sent from Reviewable.io -->

fbertsch commented:
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[mozaggregator/db.py, line 99 at r4](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_DZgNiNjzajRUrH8b2:-K_Ekk9QlyxDcltwTU_h:b-kyfk8p) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/6c8f76af4b3321bedfef4079f1ef6029c37dbb32/mozaggregator/db.py#L99)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

Which condition does this `else` clause belong to? 
</blockquote></details>

The for loop. If it doesn't break out of the loop, the else is called.

---

*[tests/dataset.py, line 112 at r4](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_DeSj2aQXhH74TZM_h:-K_EkzBHr8Ha6UEUc4Eh:btr2cbb) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/6c8f76af4b3321bedfef4079f1ef6029c37dbb32/tests/dataset.py#L112)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

Are you sure this is the right format? I don't see nested scalars in the raw pings. Maybe @dexterp37 can weigh in. It would also be useful to put a reference to the scalars format in a comment. 
</blockquote></details>

These were some that I found from actual pings, but I would definitely like to get some verification if you're unsure.

---

*[tests/test_db.py, line 245 at r4](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_Dg9QdA_XnyPLlDFhs:-K_EpDxrfNhswi9UiCdx:b-b4dq31) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/6c8f76af4b3321bedfef4079f1ef6029c37dbb32/tests/test_db.py#L245)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

Can `test_simple_measure` and `test_numeric_scalar` be refactored into the same function?
</blockquote></details>

Done. Though it has a few more arguments than I'd normally like in a function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22)*
<!-- Sent from Reviewable.io -->

Dexterp37 commented:
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[tests/dataset.py, line 112 at r4](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_DeSj2aQXhH74TZM_h:-K_EuI5Q8eb1HIzf-k8A:b4qt63l) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/6c8f76af4b3321bedfef4079f1ef6029c37dbb32/tests/dataset.py#L112)):*
<details><summary><i>Previously, fbertsch (Frank Bertsch) wrote…</i></summary><blockquote>

These were some that I found from actual pings, but I would definitely like to get some verification if you're unsure.
</blockquote></details>

Mh, the dashes look suspicious, we should definitely send nothing with a dash. The correct format is the same as the one documented [here](https://gecko.readthedocs.io/en/latest/toolkit/components/telemetry/telemetry/collection/scalars.html#the-yaml-definition-file): basically alpha numeric plus _ (underscore) and . (dot) as separators.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22)*
<!-- Sent from Reviewable.io -->

fbertsch commented:
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[tests/dataset.py, line 112 at r4](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_DeSj2aQXhH74TZM_h:-K_ExEiKO3kUTPRdG2Sz:bsd863q) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/6c8f76af4b3321bedfef4079f1ef6029c37dbb32/tests/dataset.py#L112)):*
<details><summary><i>Previously, Dexterp37 (Alessio Placitelli) wrote…</i></summary><blockquote>

Mh, the dashes look suspicious, we should definitely send nothing with a dash. The correct format is the same as the one documented [here](https://gecko.readthedocs.io/en/latest/toolkit/components/telemetry/telemetry/collection/scalars.html#the-yaml-definition-file): basically alpha numeric plus _ (underscore) and . (dot) as separators.
</blockquote></details>

@dexterp37, what about the nested scalars?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22)*
<!-- Sent from Reviewable.io -->

Dexterp37 commented:
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[tests/dataset.py, line 112 at r4](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_DeSj2aQXhH74TZM_h:-K_EyZ2wjVWVJKq8tRt6:b-okn1hc) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/6c8f76af4b3321bedfef4079f1ef6029c37dbb32/tests/dataset.py#L112)):*
<details><summary><i>Previously, fbertsch (Frank Bertsch) wrote…</i></summary><blockquote>

@dexterp37, what about the nested scalars?
</blockquote></details>

That's a sample from my machine. As you can see, keyed scalars (nested) follow the same rules:

        "scalars": {
          "browser.engagement.max_concurrent_tab_count": 70,
          "browser.engagement.max_concurrent_window_count": 1,
        },
        "keyedScalars": {
          "browser.engagement.navigation.searchbar": {
            "search_enter": 7
          }
        }

The only restrictions on the key format ("search_enter" in this case) are documented [here](https://gecko.readthedocs.io/en/latest/toolkit/components/telemetry/telemetry/collection/scalars.html#string-type-restrictions)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22)*
<!-- Sent from Reviewable.io -->

vitillo commented:
Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[tests/dataset.py, line 112 at r4](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_DeSj2aQXhH74TZM_h:-K_F66tKW2mKABLILOgX:ba213w7) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/6c8f76af4b3321bedfef4079f1ef6029c37dbb32/tests/dataset.py#L112)):*
<details><summary><i>Previously, Dexterp37 (Alessio Placitelli) wrote…</i></summary><blockquote>

That's a sample from my machine. As you can see, keyed scalars (nested) follow the same rules:

        "scalars": {
          "browser.engagement.max_concurrent_tab_count": 70,
          "browser.engagement.max_concurrent_window_count": 1,
        },
        "keyedScalars": {
          "browser.engagement.navigation.searchbar": {
            "search_enter": 7
          }
        }

The only restrictions on the key format ("search_enter" in this case) are documented [here](https://gecko.readthedocs.io/en/latest/toolkit/components/telemetry/telemetry/collection/scalars.html#string-type-restrictions)
</blockquote></details>

Right, so it seems that keyed scalars aren't handled at all in this PR. 

Frank, can you reproduce the dashes bug or was it a typo on your side? I also don't see how you could get that nested scalar in the "scalars" section to begin with.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22)*
<!-- Sent from Reviewable.io -->

fbertsch commented:
Review status: 5 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[tests/dataset.py, line 112 at r4](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_DeSj2aQXhH74TZM_h:-K_FZEFiP1D9jgGuwHqf:bec1mp2) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/6c8f76af4b3321bedfef4079f1ef6029c37dbb32/tests/dataset.py#L112)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

Right, so it seems that keyed scalars aren't handled at all in this PR. 

Frank, can you reproduce the dashes bug or was it a typo on your side? I also don't see how you could get that nested scalar in the "scalars" section to begin with.
</blockquote></details>

Must have been a typo on my end, I put that in there two months ago so I don't remember the sequence of events. Looking now, there are no dashes in about:telemetry, so it makes sense.

Do we want to add keyed scalars to this PR? Makes sense to me, if everything is already implemented.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22)*
<!-- Sent from Reviewable.io -->

fbertsch commented:
@vitillo, implemented support for keyed scalars, and made the changes so that the current numeric scalars are correct.

Dexterp37 commented:
@fbertsch IMHO there's an additional problem that needs to be addressed before landing this: we need to blacklist some keyed scalars too as we do for the [SEARCH_COUNTS](https://github.com/mozilla/python_mozaggregator/blob/4e8aaabbb51e5366f651564087b304be05daa9a9/mozaggregator/service.py#L165) histogram. I think the list should contain all the scalars in the "browser.engagement.navigation*" group. @georgf what's your opinion on this?

---

Review status: 2 of 7 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-:-K_TKNTMyGTIelm2AFuj:b42n2u5)*
<!-- Sent from Reviewable.io -->

vitillo commented:
Reviewed 1 of 5 files at r4.
Review status: 3 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[mozaggregator/aggregator.py, line 219 at r5](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_V6ix8EOfTo5LUlSh9:-K_V6ix8EOfTo5LUlShA:b-nclasi) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/83feef5691fc0b6adfa3e150f070b6f043ce67d2/mozaggregator/aggregator.py#L219)):*
> ```Python
>             for sub_name, sub_value in value.iteritems():
>                 if isinstance(sub_value, (int, float, long)):
>                     _extract_scalar_value(state, "_".join((name_prefix, name.upper(), sub_name.upper())), u"", sub_value, labels, is_child)
> ```

This looks wrong. Please have a look at how [keyed histograms](https://github.com/mozilla/python_mozaggregator/blob/master/tests/test_db.py#L236) are queried. The name is the one defined in `Histograms.json` while the **label** corresponds to the **key** of the histogram.

Nested simple measures and keyed scalars should not be handled the same way. Nested simple measures are "special snowflakes" as we had to use a backward compatible naming convention. Keyed scalars should be handled the same way keyed histograms are.

Also, it would be helpful if you filed two commits: one for numeric scalars and the other for keyed ones.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22)*
<!-- Sent from Reviewable.io -->

georgf commented:
> we need to blacklist some keyed scalars too as we do for the SEARCH_COUNTS histogram.
> I think the list should contain all the scalars in the "browser.engagement.navigation*" group.
> @georgf what's your opinion on this?

I agree, business development requires us to keep those numbers hidden.
Blacklisting all scalars in that category is easiest for now.

fbertsch commented:
@Dexterp37, @georgf - thanks for the feedback. I went ahead and disallowed any scalars starting with "browser.engagement.navigation". 

@vitillo, I split the commits up, and implemented the keyed scalars as the keyed histograms are (with appropriate labels). RFAL. Thanks for sticking with this never-ending review :)

vitillo commented:
Reviewed 4 of 4 files at r6.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[mozaggregator/aggregator.py, line 17 at r6](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_inKOTJ0y-GuiERrbb:-K_inKOUpHolTL2pf46V:bn5yxyu) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/b961f0a7a9956f346e9cff1ad8e951b0a53d66aa/mozaggregator/aggregator.py#L17)):*
> ```Python
> simple_measures_labels = cached_exponential_buckets(1, 30000, 50)
> count_histogram_labels = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 21, 23, 25, 27, 29, 31, 34, 37, 40, 43, 46, 50, 54, 58, 63, 68, 74, 80, 86, 93, 101, 109, 118, 128, 138, 149, 161, 174, 188, 203, 219, 237, 256, 277, 299, 323, 349, 377, 408, 441, 477, 516, 558, 603, 652, 705, 762, 824, 891, 963, 1041, 1125, 1216, 1315, 1422, 1537, 1662, 1797, 1943, 2101, 2271, 2455, 2654, 2869, 3102, 3354, 3626, 3920, 4238, 4582, 4954, 5356, 5791, 6261, 6769, 7318, 7912, 8554, 9249, 10000]
> numeric_scalars_labels = list(count_histogram_labels)
> ```

`count_histogram_labels` is already a list

---

*[mozaggregator/aggregator.py, line 215 at r6](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_iovtOB3ehYvpXaGFQ:-K_iovtOB3ehYvpXaGFR:b-o3zlhm) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/b961f0a7a9956f346e9cff1ad8e951b0a53d66aa/mozaggregator/aggregator.py#L215)):*
> ```Python
>             _extract_scalar_value(state, u"_".join((simple_measures_prefix, name.upper())), u"", value, simple_measures_labels, is_child)
> 
> def _extract_numeric_scalars(state, scalar_dict, is_child=False):
> ```

Please split this function up in two functions: one for numeric scalars and the other for keyed numeric ones.

---

*[mozaggregator/service.py, line 18 at r6](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_ipXZPzi6N1ELztgvk:-K_ipXZPzi6N1ELztgvl:b64rz39) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/b961f0a7a9956f346e9cff1ad8e951b0a53d66aa/mozaggregator/service.py#L18)):*
> ```Python
> from db import get_db_connection_string, histogram_revision_map
> 
> scalar_measure_map = {simple_measures_prefix: simple_measures_labels, 
> ```

This dict is redefined a couple of times. Can we have the prefixes and the dicts all in one place? In fact we might just store the dict somewhere and use that whenever either the labels or the prefixes are needed.

---

*[tests/dataset.py, line 167 at r6](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_ireIcQp5eJlWZzB43:-K_ireIcQp5eJlWZzB44:byn6khb) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/b961f0a7a9956f346e9cff1ad8e951b0a53d66aa/tests/dataset.py#L167)):*
> ```Python
> 
>     if aggregated_child_histograms:
>         processes_payload[u"content"] = {
> ```

What about scalars defined in the content process? We should support those as well I think.

---

*[tests/test_db.py, line 250 at r6](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_iuyYWrFkDUMox4ytB:-K_iuyYWrFkDUMox4ytC:bc6ai3c) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/b961f0a7a9956f346e9cff1ad8e951b0a53d66aa/tests/test_db.py#L250)):*
> ```Python
> 
>     for res in reply["data"]:
>         assert(res["count"] == expected_count * (1 if not is_child else NUM_CHILDREN_PER_PING + 1))
> ```

Please write a comment on how the count is determined to simplify maintenance in the future.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22)*
<!-- Sent from Reviewable.io -->

fbertsch commented:
Review status: 2 of 7 files reviewed at latest revision, 7 unresolved discussions.

---

*[mozaggregator/aggregator.py, line 219 at r5](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_V6ix8EOfTo5LUlSh9:-K_ix3mK3WXBsnigQ1bU:b-896fix) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/83feef5691fc0b6adfa3e150f070b6f043ce67d2/mozaggregator/aggregator.py#L219)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

This looks wrong. Please have a look at how [keyed histograms](https://github.com/mozilla/python_mozaggregator/blob/master/tests/test_db.py#L236) are queried. The name is the one defined in `Histograms.json` while the **label** corresponds to the **key** of the histogram.

Nested simple measures and keyed scalars should not be handled the same way. Nested simple measures are "special snowflakes" as we had to use a backward compatible naming convention. Keyed scalars should be handled the same way keyed histograms are.

Also, it would be helpful if you filed two commits: one for numeric scalars and the other for keyed ones.
</blockquote></details>

Done.

---

*[mozaggregator/aggregator.py, line 17 at r6](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_inKOTJ0y-GuiERrbb:-K_iwm5wnd9hfbLjwL6a:bidy7zk) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/b961f0a7a9956f346e9cff1ad8e951b0a53d66aa/mozaggregator/aggregator.py#L17)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

`count_histogram_labels` is already a list
</blockquote></details>

I did this originally in case there were any changes made to count_histogram_labels (in execution), that numeric_scalars_labels would be pointing to a different object.

---

*[mozaggregator/aggregator.py, line 215 at r6](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_iovtOB3ehYvpXaGFQ:-K_jn6-8mFOX3yNmGSVm:b-896fix) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/b961f0a7a9956f346e9cff1ad8e951b0a53d66aa/mozaggregator/aggregator.py#L215)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

Please split this function up in two functions: one for numeric scalars and the other for keyed numeric ones.
</blockquote></details>

Done.

---

*[mozaggregator/service.py, line 18 at r6](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_ipXZPzi6N1ELztgvk:-K_jvlIKh2sHzoLOsxsv:bp2twzu) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/b961f0a7a9956f346e9cff1ad8e951b0a53d66aa/mozaggregator/service.py#L18)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

This dict is redefined a couple of times. Can we have the prefixes and the dicts all in one place? In fact we might just store the dict somewhere and use that whenever either the labels or the prefixes are needed.
</blockquote></details>

I defined the dict globally, but I kept the labels and prefixes as well. Getting them all from a dict didn't make the code any simpler (since there are only three, and presumably won't be any more).

---

*[tests/dataset.py, line 167 at r6](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_ireIcQp5eJlWZzB43:-K_jvcwZbbKXKA-qsght:b-4ib0os) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/b961f0a7a9956f346e9cff1ad8e951b0a53d66aa/tests/dataset.py#L167)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

What about scalars defined in the content process? We should support those as well I think.
</blockquote></details>

I made Bug 1319610 for this. Content process scalars are not ready yet.

---

*[tests/test_db.py, line 250 at r6](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_iuyYWrFkDUMox4ytB:-K_kXhHJZV-JnaFtRWim:b-896fix) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/b961f0a7a9956f346e9cff1ad8e951b0a53d66aa/tests/test_db.py#L250)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

Please write a comment on how the count is determined to simplify maintenance in the future.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22)*
<!-- Sent from Reviewable.io -->

fbertsch commented:
@vitillo, RFAL

fbertsch commented:
@vitillo, Added scalar labels.

vitillo commented:
Reviewed 7 of 7 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[mozaggregator/aggregator.py, line 17 at r6](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_inKOTJ0y-GuiERrbb:-Ka19QRaDhIBW9AiAAi_:b-im3wjm) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/b961f0a7a9956f346e9cff1ad8e951b0a53d66aa/mozaggregator/aggregator.py#L17)):*
<details><summary><i>Previously, fbertsch (Frank Bertsch) wrote…</i></summary><blockquote>

I did this originally in case there were any changes made to count_histogram_labels (in execution), that numeric_scalars_labels would be pointing to a different object.
</blockquote></details>

Let's consider those fields as [constants](https://www.python.org/dev/peps/pep-0008/#constants).

---

*[mozaggregator/aggregator.py, line 221 at r7](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka1A7HgNBkL9nC6i_aj:-Ka1A7HgNBkL9nC6i_ak:b-4p13v3) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/7aec192cae2ef90b7a841c1e051cd7f27bc50f11/mozaggregator/aggregator.py#L221)):*
> ```Python
>             _extract_scalar_value(state, u"_".join((simple_measures_prefix, name.upper())), u"", value, simple_measures_labels, is_child)
> 
> def _extract_numeric_scalars(state, scalar_dict, is_child=False):
> ```

This function accepts `is_child` as parameter, yet content scalars aren't supported yet. Please either implement the required functionality to support content histograms or remove references to them. The same applies in other parts of this PR.

---

*[mozaggregator/service.py, line 175 at r7](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka1CugpAScAWJ_ci6Va:-Ka1CugpAScAWJ_ci6Vb:bywe2kt) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/7aec192cae2ef90b7a841c1e051cd7f27bc50f11/mozaggregator/service.py#L175)):*
> ```Python
>     return new_defs
> 
> def _get_description(channel, prefix, metric):
> ```

We might need more than the description in the future. Please create a class that nicely wraps the definition of a [keyed] scalar. You might want to move this code to its own module as well. 

It might also make sense to consider adding this code to `python_moztelemetry` in a follow up Bug so that our analysts can reuse it to fetch scalar definitions.

---

*[tests/test_db.py, line 250 at r6](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_iuyYWrFkDUMox4ytB:-Ka1Da1cpfCcN9jdEKjx:b-3lhdqo) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/b961f0a7a9956f346e9cff1ad8e951b0a53d66aa/tests/test_db.py#L250)):*
<details><summary><i>Previously, fbertsch (Frank Bertsch) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Hm, I don't see the comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22)*
<!-- Sent from Reviewable.io -->

fbertsch commented:
Review status: 3 of 8 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[mozaggregator/aggregator.py, line 17 at r6](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_inKOTJ0y-GuiERrbb:-Ka2r2890BE2iEN8St7-:b-896fix) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/b961f0a7a9956f346e9cff1ad8e951b0a53d66aa/mozaggregator/aggregator.py#L17)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

Let's consider those fields as [constants](https://www.python.org/dev/peps/pep-0008/#constants).
</blockquote></details>

Done.

---

*[mozaggregator/aggregator.py, line 221 at r7](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka1A7HgNBkL9nC6i_aj:-Ka2r10dUEYgpENVFPQF:b-896fix) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/7aec192cae2ef90b7a841c1e051cd7f27bc50f11/mozaggregator/aggregator.py#L221)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

This function accepts `is_child` as parameter, yet content scalars aren't supported yet. Please either implement the required functionality to support content histograms or remove references to them. The same applies in other parts of this PR.
</blockquote></details>

Done.

---

*[mozaggregator/service.py, line 175 at r7](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka1CugpAScAWJ_ci6Va:-Ka2r7JZNb-sM2L0Bw56:b-i6e6z6) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/7aec192cae2ef90b7a841c1e051cd7f27bc50f11/mozaggregator/service.py#L175)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

We might need more than the description in the future. Please create a class that nicely wraps the definition of a [keyed] scalar. You might want to move this code to its own module as well. 

It might also make sense to consider adding this code to `python_moztelemetry` in a follow up Bug so that our analysts can reuse it to fetch scalar definitions.
</blockquote></details>

Added here, and I made bug 

---

*[tests/test_db.py, line 250 at r6](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_iuyYWrFkDUMox4ytB:-Ka3h1x2oWe0A78sw-6G:bcu8ryp) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/b961f0a7a9956f346e9cff1ad8e951b0a53d66aa/tests/test_db.py#L250)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

Hm, I don't see the comment.
</blockquote></details>

Hmm, must have been lost in the git shuffling I did. It's been added.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22)*
<!-- Sent from Reviewable.io -->

fbertsch commented:
Review status: 3 of 8 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[mozaggregator/service.py, line 175 at r7](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka1CugpAScAWJ_ci6Va:-Ka3jmP7hAoBrBJLUU23:bcrne3p) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/7aec192cae2ef90b7a841c1e051cd7f27bc50f11/mozaggregator/service.py#L175)):*
<details><summary><i>Previously, fbertsch (Frank Bertsch) wrote…</i></summary><blockquote>

Added here, and I made bug 
</blockquote></details>

Whoops, meant to say I commented on bug 1297080, which this applies to.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22)*
<!-- Sent from Reviewable.io -->

vitillo commented:
Reviewed 5 of 5 files at r8.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[mozaggregator/aggregator.py, line 17 at r6](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_inKOTJ0y-GuiERrbb:-Ka7jaD6DYQIoQ_QDTxw:b-bd3c8x) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/b961f0a7a9956f346e9cff1ad8e951b0a53d66aa/mozaggregator/aggregator.py#L17)):*
<details><summary><i>Previously, fbertsch (Frank Bertsch) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Can you please rename them to upper case to reflect that?

---

*[mozaggregator/scalar.py, line 16 at r8](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka7jt0opjFfDQ-YGExf:-Ka7jt0pP1NK4Jv3j1A8:beo6m06) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/f0916c1ccbd0d61cf495e5049f010129c649ff7e/mozaggregator/scalar.py#L16)):*
> ```Python
> 
> 
> def _yaml_unnest(defs, prefix=''):
> ```

Can we make this a static method of the Scalar class? 

---

*[mozaggregator/scalar.py, line 23 at r8](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka7kK_iidX_4mxX1QWz:-Ka7kK_iidX_4mxX1QX-:b3wvnnm) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/f0916c1ccbd0d61cf495e5049f010129c649ff7e/mozaggregator/scalar.py#L23)):*
> ```Python
>     # prints {'browser.nav.clicks': {'description': 'a description', 'expires': 'never'}}
>     """
>     stop = lambda x: type(x) is not dict or x.viewkeys() & Scalar.REQUIRED_FIELDS
> ```

Please run this file through flake8.

---

*[mozaggregator/scalar.py, line 36 at r8](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka7jy0d4dt0piyNE3np:-Ka7jy0d4dt0piyNE3nq:bisxrxg) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/f0916c1ccbd0d61cf495e5049f010129c649ff7e/mozaggregator/scalar.py#L36)):*
> ```Python
> 
> 
> def _get_scalar_definition(url, metric, additional_scalars=None):
> ```

Can we make this a static method of the Scalar class? 

---

*[mozaggregator/scalar.py, line 40 at r8](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka7n2MAEdNK_pGl1SuG:-Ka7n2MAEdNK_pGl1SuH:bxjy0ol) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/f0916c1ccbd0d61cf495e5049f010129c649ff7e/mozaggregator/scalar.py#L40)):*
> ```Python
>         content = requests.get(url).content
>         definitions = _yaml_unnest(yaml.load(content))
>         definition_cache[url] = json.dumps(definitions)
> ```

Is it necessary to serialize/deserialize to/from JSON?

---

*[mozaggregator/scalar.py, line 44 at r8](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka7nLDt4DRa5BnuBf1G:-Ka7nLDt4DRa5BnuBf1H:b-6zewwt) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/f0916c1ccbd0d61cf495e5049f010129c649ff7e/mozaggregator/scalar.py#L44)):*
> ```Python
>         definitions = json.loads(definition_cache[url])
> 
>     if additional_scalars:
> ```

Please remove this as there are no use cases for now afaik.

---

*[mozaggregator/scalar.py, line 50 at r8](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka7k21dXDmowJGekUFE:-Ka7k21dXDmowJGekUFF:b-yfmcqn) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/f0916c1ccbd0d61cf495e5049f010129c649ff7e/mozaggregator/scalar.py#L50)):*
> ```Python
> 
> 
> class Scalar:
> ```

Please add some tests for the `Scalar` class.

---

*[mozaggregator/scalar.py, line 63 at r8](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka7ncWKQq1NmK5hjRMX:-Ka7ncWKQq1NmK5hjRMY:b-tw80cv) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/f0916c1ccbd0d61cf495e5049f010129c649ff7e/mozaggregator/scalar.py#L63)):*
> ```Python
>         Initialize a scalar from it's name.
> 
>         :param channel: The channel for which the scalar is defined. One of
> ```

Please document all parameters.

---

*[mozaggregator/scalar.py, line 74 at r8](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka7oAgAhsuQ4q7ts2lL:-Ka7oAgB9Er2NYTCmxe5:b-i48sn9) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/f0916c1ccbd0d61cf495e5049f010129c649ff7e/mozaggregator/scalar.py#L74)):*
> ```Python
> 
>         if picked == 0:
>             channel = 'nightly'
> ```

Please document this default.

---

*[mozaggregator/scalar.py, line 103 at r8](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka7oSH-7t86aiAhjz2z:-Ka7oSH-7t86aiAhjz3-:bltnugw) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/f0916c1ccbd0d61cf495e5049f010129c649ff7e/mozaggregator/scalar.py#L103)):*
> ```Python
>         return self.definition
> 
>     def get(self, arg, default=None):
> ```

Please remove this method as it's doing too many things.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22)*
<!-- Sent from Reviewable.io -->

fbertsch commented:
Review status: 1 of 9 files reviewed at latest revision, 11 unresolved discussions.

---

*[mozaggregator/aggregator.py, line 17 at r6](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_inKOTJ0y-GuiERrbb:-KaAcNyUcCbZmMnDpLxM:b-896fix) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/b961f0a7a9956f346e9cff1ad8e951b0a53d66aa/mozaggregator/aggregator.py#L17)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

Can you please rename them to upper case to reflect that?
</blockquote></details>

Done.

---

*[mozaggregator/scalar.py, line 16 at r8](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka7jt0opjFfDQ-YGExf:-KaAcP22bPlXatmVLjx7:b-896fix) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/f0916c1ccbd0d61cf495e5049f010129c649ff7e/mozaggregator/scalar.py#L16)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

Can we make this a static method of the Scalar class? 
</blockquote></details>

Done.

---

*[mozaggregator/scalar.py, line 23 at r8](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka7kK_iidX_4mxX1QWz:-KaAcPjNT3sZy49YoFg-:bes7rct) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/f0916c1ccbd0d61cf495e5049f010129c649ff7e/mozaggregator/scalar.py#L23)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

Please run this file through flake8.
</blockquote></details>

Flake8 was not happy! Fixed.

---

*[mozaggregator/scalar.py, line 36 at r8](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka7jy0d4dt0piyNE3np:-KaAcSzDPP_ZOHLC7S2r:b-896fix) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/f0916c1ccbd0d61cf495e5049f010129c649ff7e/mozaggregator/scalar.py#L36)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

Can we make this a static method of the Scalar class? 
</blockquote></details>

Done.

---

*[mozaggregator/scalar.py, line 40 at r8](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka7n2MAEdNK_pGl1SuG:-KaAcTT3WEuWya_WJ6uv:bcf5hlg) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/f0916c1ccbd0d61cf495e5049f010129c649ff7e/mozaggregator/scalar.py#L40)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

Is it necessary to serialize/deserialize to/from JSON?
</blockquote></details>

Nope! Fixed.

---

*[mozaggregator/scalar.py, line 44 at r8](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka7nLDt4DRa5BnuBf1G:-KaAcY2Vkgk9VkEigoYb:b-896fix) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/f0916c1ccbd0d61cf495e5049f010129c649ff7e/mozaggregator/scalar.py#L44)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

Please remove this as there are no use cases for now afaik.
</blockquote></details>

Done.

---

*[mozaggregator/scalar.py, line 50 at r8](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka7k21dXDmowJGekUFE:-KaAcYe1r2jS8682Oriz:b-896fix) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/f0916c1ccbd0d61cf495e5049f010129c649ff7e/mozaggregator/scalar.py#L50)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

Please add some tests for the `Scalar` class.
</blockquote></details>

Done.

---

*[mozaggregator/scalar.py, line 63 at r8](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka7ncWKQq1NmK5hjRMX:-KaAcZQZs-MM8JEaaVYI:b-896fix) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/f0916c1ccbd0d61cf495e5049f010129c649ff7e/mozaggregator/scalar.py#L63)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

Please document all parameters.
</blockquote></details>

Done.

---

*[mozaggregator/scalar.py, line 74 at r8](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka7oAgAhsuQ4q7ts2lL:-KaAc_4Hb_w0vf_nCx1G:b-896fix) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/f0916c1ccbd0d61cf495e5049f010129c649ff7e/mozaggregator/scalar.py#L74)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

Please document this default.
</blockquote></details>

Done.

---

*[mozaggregator/scalar.py, line 103 at r8](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-Ka7oSH-7t86aiAhjz2z:-KaAc_gK3cuOcv3T4cZq:b-896fix) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/f0916c1ccbd0d61cf495e5049f010129c649ff7e/mozaggregator/scalar.py#L103)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

Please remove this method as it's doing too many things.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22)*
<!-- Sent from Reviewable.io -->

vitillo commented:
Reviewed 7 of 8 files at r9, 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[mozaggregator/scalar.py, line 14 at r10](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-KaCa-0YHwhD0pHw8yUr:-KaCa-0ZrXObF4HLhoHD:b2ylapg) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/ea76fb3a08c7b7f9a366c26199efa5464f45904c/mozaggregator/scalar.py#L14)):*
> ```Python
>              'release': 'https://hg.mozilla.org/releases/mozilla-release/rev/tip'}
> 
> definition_cache = ExpiringDict(max_len=2**10, max_age_seconds=3600)
> ```

This should be a static field of the `Scalar` class.

---

*[mozaggregator/scalar.py, line 57 at r10](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-KaCaPZkRboHJcHNMGVs:-KaCaPZlQEd0EDLNKX9P:b-yis82v) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/ea76fb3a08c7b7f9a366c26199efa5464f45904c/mozaggregator/scalar.py#L57)):*
> ```Python
>         self.name = name
>         self.value = value
>         self.definition = json_definition
> ```

s/json_definition/definition

---

*[tests/test_scalar.py, line 12 at r10](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-KaCagMPTCV55tfirbZh:-KaCagMPTCV55tfirbZi:b-vnhgvb) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/ea76fb3a08c7b7f9a366c26199efa5464f45904c/tests/test_scalar.py#L12)):*
> ```Python
>     assert scalar.get_name() == name
>     assert scalar.get_definition().get('kind') == 'uint'
>     assert Scalar.REQUIRED_FIELDS.issubset(scalar.get_definition().viewkeys())
> ```

Please run this through flake8. Might be worthwhile filing a follow up bug to verify flake8 compliance on push with travis.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22)*
<!-- Sent from Reviewable.io -->

fbertsch commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[mozaggregator/scalar.py, line 14 at r10](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-KaCa-0YHwhD0pHw8yUr:-KaCwn9imf4mJW2QctE_:b-896fix) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/ea76fb3a08c7b7f9a366c26199efa5464f45904c/mozaggregator/scalar.py#L14)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

This should be a static field of the `Scalar` class.
</blockquote></details>

Done.

---

*[mozaggregator/scalar.py, line 57 at r10](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-KaCaPZkRboHJcHNMGVs:-KaCwn_jqtI7iiPpin64:b-896fix) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/ea76fb3a08c7b7f9a366c26199efa5464f45904c/mozaggregator/scalar.py#L57)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

s/json_definition/definition
</blockquote></details>

Done.

---

*[tests/test_scalar.py, line 12 at r10](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-KaCagMPTCV55tfirbZh:-KaCwrf-c9oIhke7DCit:buobg43) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/ea76fb3a08c7b7f9a366c26199efa5464f45904c/tests/test_scalar.py#L12)):*
<details><summary><i>Previously, vitillo (Roberto Agostino Vitillo) wrote…</i></summary><blockquote>

Please run this through flake8. Might be worthwhile filing a follow up bug to verify flake8 compliance on push with travis.
</blockquote></details>

I'm fine with flake8 compliance, but with some tweaks. The 80 character line count limitation, for example, should be allowed in some cases. What we *should* have is a flake8/pylint config that we use globally across all of our Python projects.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22)*
<!-- Sent from Reviewable.io -->

vitillo commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 3 files at r11, 1 of 1 files at r12.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-:-KaCydTlPmG6xYQN_PQK:bnfp4nl)*
<!-- Sent from Reviewable.io -->

fbertsch commented:
WOOH! Let's get this merged after the build passes.

fbertsch commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[tests/dataset.py, line 112 at r4](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22#-K_DeSj2aQXhH74TZM_h:-KaCzdxKIbKKlUcZJ231:b-896fix) ([raw file](https://github.com/mozilla/python_mozaggregator/blob/6c8f76af4b3321bedfef4079f1ef6029c37dbb32/tests/dataset.py#L112)):*
<details><summary><i>Previously, fbertsch (Frank Bertsch) wrote…</i></summary><blockquote>

Must have been a typo on my end, I put that in there two months ago so I don't remember the sequence of events. Looking now, there are no dashes in about:telemetry, so it makes sense.

Do we want to add keyed scalars to this PR? Makes sense to me, if everything is already implemented.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mozilla/python_mozaggregator/22)*
<!-- Sent from Reviewable.io -->
